File: frontend/src/lib/tool-view.test.ts
import { describe, expect, it } from "vitest"

import { execCommand, summariseToolCall, toolRowSummary, viewTool } from "./tool-view"

describe("summariseToolCall", () => {
  it("shows the command, not the JSON envelope", () => {
    expect(summariseToolCall("exec", `{"command":"echo hi","cwd":"."}`)).toBe("echo hi")
    expect(summariseToolCall("exec", `{"command":"echo hi"}`)).not.toMatch(/[{"]/)
    expect(summariseToolCall("exec", `{"cwd":"."}`)).toBe("")
  })

  it("shows the search query, not the JSON envelope", () => {
    expect(summariseToolCall("web_search", `{"query":"alpha terms"}`)).toBe("alpha terms")
    expect(summariseToolCall("web_search", `{"query":"alpha terms"}`)).not.toContain("query")
  })

  it("joins a pattern with its path when both are present", () => {
    expect(summariseToolCall("grep", `{"pattern":"TODO","path":"src"}`)).toBe("TODO · src")
    expect(summariseToolCall("wait_agents", `{"timeout_s":60,"agent_ids":["a-1"]}`)).toBe("")
  })

  it("falls back to flattened text when args are not JSON", () => {
    expect(summariseToolCall("exec", "echo hi")).toBe("echo hi")
  })

  it("flattens a multiline command for the one-line summary", () => {
    expect(summariseToolCall("exec", JSON.stringify({ command: "echo hi\n&& ls" }))).toBe(
      "echo hi && ls",
    )
  })
})

describe("execCommand", () => {
  it("keeps newlines so expand can show the whole invocation", () => {
    expect(execCommand(JSON.stringify({ command: "cat <<END\nline\nEND", cwd: "." }))).toBe(
      "cat <<END\nline\nEND",
    )
  })

  it("is empty when there is no command", () => {
    expect(execCommand(`{"cwd":"."}`)).toBe("")
    expect(execCommand("")).toBe("")
  })
})

describe("viewTool", () => {
  it("surfaces a failed command in red-ready fields, not as pretty JSON", () => {
    const raw = JSON.stringify({
      exit_code: 127,
      stdout: "",
      stderr: "not found",
      failed: true,
      error: "exit status 127",
      full_command: "nope",
    })
    const view = viewTool("exec", `{"command":"nope"}`, raw)
    expect(view.summary).toBe("nope")
    expect(view.failed).toBe(true)
    expect(view.error).toBe("exit status 127")
    expect(view.body).toBe("not found")
    expect(view.body).not.toContain("full_command")
    expect(view.body).not.toContain("{")
  })

  it("shows stdout for a successful command", () => {
    const raw = JSON.stringify({ exit_code: 0, stdout: "ok\nline", stderr: "", failed: false })
    const view = viewTool("exec", `{"command":"echo ok"}`, raw)
    expect(view.failed).toBe(false)
    expect(view.body).toBe("ok\nline")
  })

  it("lists search hits instead of dumping JSON", () => {
    const raw = JSON.stringify({
      message: "ok",
      results: [
        { title: "One", url: "https://example.invalid/one", summary: "first hit" },
        { title: "Two", url: "https://example.invalid/two" },
      ],
    })
    const view = viewTool("web_search", `{"query":"alpha terms"}`, raw)
    expect(view.hits).toHaveLength(2)
    expect(view.hits?.[0]).toEqual({
      title: "One",
      url: "https://example.invalid/one",
      snippet: "first hit",
    })
    expect(view.body).toBe("")
  })

  it("treats an error: prefix as a failed call", () => {
    const view = viewTool("web_fetch", `{"url":"https://example.invalid"}`, "error: timed out")
    expect(view.failed).toBe(true)
    expect(view.error).toBe("timed out")
  })

  it("shows a memory write as the action and the note, not the JSON", () => {
    expect(summariseToolCall("memory", `{"action":"add","content":"User prefers terse replies"}`)).toBe(
      "add · User prefers terse replies",
    )
    const view = viewTool(
      "memory",
      `{"action":"add","content":"User prefers terse replies"}`,
      JSON.stringify({ success: true, changed: true, usage: "24/2200", entries: ["User prefers terse replies"] }),
    )
    expect(view.failed).toBe(false)
    expect(view.body).toContain("User prefers terse replies")
    expect(view.body).toContain("24/2200")
  })

  it("flags a refused memory write and lists what is already stored", () => {
    const view = viewTool(
      "memory",
      `{"action":"add","content":"another"}`,
      JSON.stringify({
        success: false,
        error: "memory is at 30/30 characters",
        current_entries: ["the one that is already there"],
      }),
    )
    expect(view.failed).toBe(true)
    expect(view.error).toContain("30/30")
    expect(view.body).toBe("the one that is already there")
  })

  it("shows a missing skill once, and lists the ones that exist", () => {
    const view = viewTool(
      "skill_view",
      `{"name":"absent"}`,
      JSON.stringify({
        success: false,
        error: 'no skill named "absent". skill_view only opens skills recorded in this project\'s memory',
        available: ["present"],
      }),
    )
    expect(view.failed).toBe(true)
    expect(view.error).toContain("absent")
    expect(view.body).toBe("present")
    expect(view.body).not.toContain("no skill named")
  })

  it("does not repeat a missing-skill error when nothing is recorded", () => {
    const view = viewTool(
      "skill_view",
      `{"name":"absent"}`,
      JSON.stringify({
        success: false,
        error: 'no skill named "absent". skill_view only opens skills recorded in this project\'s memory',
        available: [],
      }),
    )
    expect(view.failed).toBe(true)
    expect(view.error).toContain("absent")
    expect(view.body).toBe("")
  })
})

describe("toolRowSummary", () => {
  it("puts a memory refusal on the collapsed row, not just the note that did not land", () => {
    const view = viewTool(
      "memory",
      `{"action":"replace","content":"a longer status"}`,
      JSON.stringify({
        success: false,
        error: "memory is at 2200/2200 characters; this write exceeds the limit by 40. Do not retry the same write.",
        current_entries: ["the one that is already there"],
      }),
    )
    const row = toolRowSummary(view)
    expect(row).toContain("replace")
    expect(row).toContain("Do not retry")
    expect(row).toContain("2200/2200")
  })

  it("leaves a successful row as the action and the note", () => {
    const view = viewTool(
      "memory",
      `{"action":"add","content":"a durable fact"}`,
      JSON.stringify({ success: true, changed: true, usage: "14/2200", entries: ["a durable fact"] }),
    )
    expect(toolRowSummary(view)).toBe("add · a durable fact")
  })
})

describe("file change view", () => {
  it("puts plus/minus counts on an edit path, not the JSON envelope", () => {
    const view = viewTool(
      "edit",
      JSON.stringify({
        file_path: "pkg/alpha.go",
        search_block: "return 0",
        replace_block: "return 1",
      }),
      "ok: replaced block in pkg/alpha.go",
    )
    expect(view.summary).toContain("pkg/alpha.go")
    expect(view.summary).toMatch(/\+1/)
    expect(view.summary).toMatch(/−1/)
    expect(view.summary).not.toContain("search_block")
    expect(view.summary).not.toContain("{")
    expect(toolRowSummary(view)).toContain("+1")
    expect(view.diff?.added).toBe(1)
    expect(view.body).toBe("")
  })

  it("puts an added-line count on a write path, not the status sentence", () => {
    const view = viewTool(
      "write",
      JSON.stringify({
        file_path: "pkg/alpha.go",
        content: "package alpha\nfunc Alpha() {}\n",
      }),
      "Updated file pkg/alpha.go",
    )
    expect(view.summary).toContain("pkg/alpha.go")
    expect(view.summary).toMatch(/\+2/)
    expect(view.summary).not.toContain("content")
    expect(view.summary).not.toContain("{")
    expect(toolRowSummary(view)).toContain("+2")
    expect(view.diff?.added).toBe(2)
    expect(view.body).toBe("")
  })

  it("does not advertise +0 on an empty write, but still attaches the hunk", () => {
    const view = viewTool(
      "write",
      JSON.stringify({ file_path: "pkg/alpha.go", content: "" }),
      "Updated file pkg/alpha.go",
    )
    expect(view.summary).toBe("pkg/alpha.go")
    expect(view.summary).not.toMatch(/\+0/)
    expect(view.diff?.added).toBe(0)
    expect(view.diff?.hunks[0].lines).toEqual([])
  })

  it("attaches the hunk while the call is still pending", () => {
    const view = viewTool(
      "edit",
      JSON.stringify({
        file_path: "pkg/alpha.go",
        search_block: "return 0",
        replace_block: "return 1",
      }),
    )
    expect(view.diff?.added).toBe(1)
    expect(view.body).toBe("")
  })
})

File: internal/tui/toolview_test.go
package tui

import (
	"encoding/json"
	"strings"
	"testing"
)

func TestSummariseToolArgsShowsTheCommandNotTheJSON(t *testing.T) {
	got := summariseToolArgs("exec", `{"command":"echo hi","cwd":"."}`)
	if got != "echo hi" {
		t.Fatalf("exec summary = %q", got)
	}
	if strings.ContainsAny(got, `{}"`) {
		t.Fatalf("the JSON envelope leaked into the summary: %q", got)
	}
}

func TestSummariseToolArgsShowsTheSearchQuery(t *testing.T) {
	got := summariseToolArgs("web_search", `{"query":"alpha terms"}`)
	if got != "alpha terms" {
		t.Fatalf("search summary = %q", got)
	}
	if strings.Contains(got, "query") {
		t.Fatalf("the field name leaked into the summary: %q", got)
	}
}

func TestSummariseToolArgsKeepsAPlainPath(t *testing.T) {
	// Older notifications were `read(notes.md)`, not a JSON object.
	if got := summariseToolArgs("read", "notes.md"); got != "notes.md" {
		t.Fatalf("plain args = %q", got)
	}
}

func TestViewToolResultSurfacesAFailedCommand(t *testing.T) {
	raw, _ := json.Marshal(map[string]any{
		"exit_code":    127,
		"stdout":       "",
		"stderr":       "not found",
		"failed":       true,
		"error":        "exit status 127",
		"full_command": "nope",
	})
	view := viewToolResult("exec", string(raw))
	if !view.failed || view.error != "exit status 127" {
		t.Fatalf("failed view = %+v", view)
	}
	if strings.Contains(view.display(), "full_command") || strings.Contains(view.body, "{") {
		t.Fatalf("JSON envelope leaked into the result: %q", view.display())
	}
	if view.body != "not found" {
		t.Fatalf("stderr should be the body, got %q", view.body)
	}
}

func TestViewToolResultListsSearchHits(t *testing.T) {
	raw, _ := json.Marshal(map[string]any{
		"results": []map[string]any{
			{"title": "One", "url": "https://example.invalid/one", "summary": "first hit"},
		},
	})
	view := viewToolResult("web_search", string(raw))
	if view.failed || !strings.Contains(view.body, "One") || strings.Contains(view.body, `"title"`) {
		t.Fatalf("search view = %+v", view)
	}
}

func TestViewToolResultTreatsAnErrorPrefixAsFailed(t *testing.T) {
	view := viewToolResult("web_fetch", "error: timed out")
	if !view.failed || view.error != "timed out" {
		t.Fatalf("prefix view = %+v", view)
	}
}

func TestSummariseToolArgsCoversTheFallbackPaths(t *testing.T) {
	if got := summariseToolArgs("exec", ""); got != "" {
		t.Fatalf("empty args = %q", got)
	}
	if got := summariseToolArgs("grep", `{"pattern":"TODO","path":"src"}`); got != "TODO · src" {
		t.Fatalf("grep = %q", got)
	}
	if got := summariseToolArgs("unknown", `{"url":"https://example.invalid"}`); got != "https://example.invalid" {
		t.Fatalf("fallback key = %q", got)
	}
	if got := summariseToolArgs("unknown", `{"role":"researcher"}`); got != "researcher" {
		t.Fatalf("first value = %q", got)
	}
	if got := summariseToolArgs("exec", `{"cwd":"."}`); got != "" {
		t.Fatalf("exec without a command should stay empty, got %q", got)
	}
	if got := summariseToolArgs("wait_agents", `{"timeout_s":60,"agent_ids":["a-1"]}`); got != "" {
		t.Fatalf("wait_agents should not pick a stray field, got %q", got)
	}
}

func TestViewToolResultParsesStdoutAndBareExit(t *testing.T) {
	ok := viewToolResult("python_runner", `{"exit_code":0,"stdout":"ok\nline","stderr":"","failed":false}`)
	if ok.failed || ok.body != "ok\nline" {
		t.Fatalf("success view = %+v", ok)
	}
	fail := viewToolResult("exec", `{"exit_code":2,"stdout":"out","stderr":"err","failed":true}`)
	if !fail.failed || fail.error != "exit 2" || !strings.Contains(fail.display(), "out") {
		t.Fatalf("bare exit view = %+v display=%q", fail, fail.display())
	}
	plain := viewToolResult("ls", "a\nb")
	if plain.failed || plain.body != "a\nb" || plain.display() != "a\nb" {
		t.Fatalf("plain view = %+v", plain)
	}
	if viewToolResult("exec", "").display() != "" {
		t.Fatal("empty result should display nothing")
	}
	raw := viewToolResult("exec", `{"file_path":"x"}`)
	if raw.failed || !strings.Contains(raw.body, "file_path") {
		t.Fatalf("non-run exec payload = %+v", raw)
	}
	bareFail := viewToolResult("exec", `{"failed":true}`)
	if !bareFail.failed || bareFail.error != "exit ?" {
		t.Fatalf("failed without a code = %+v", bareFail)
	}
	if _, ok := parseRunResult(`{"file_path":"x"}`); ok {
		t.Fatal("a non-run payload must not parse as a command result")
	}
	if _, ok := parseRunResult(`{"stderr":"e"}`); !ok {
		t.Fatal("stderr alone is still a run payload")
	}
}

func TestParseSearchHitsAcceptsWrappedAndBareLists(t *testing.T) {
	bare, _ := json.Marshal([]map[string]any{
		{"name": "Bare", "href": "https://example.invalid/b", "description": "desc"},
		{"url": "https://example.invalid/only"},
		{"nested": true},
	})
	if hits := parseSearchHits(string(bare)); len(hits) != 2 {
		t.Fatalf("bare hits = %v", hits)
	}
	items, _ := json.Marshal(map[string]any{
		"items": []map[string]any{{"title": "Item", "link": "https://example.invalid/i", "content": "body"}},
	})
	if hits := parseSearchHits(string(items)); len(hits) != 1 || !strings.Contains(hits[0], "Item") {
		t.Fatalf("items hits = %v", hits)
	}
	data, _ := json.Marshal(map[string]any{
		"data": []map[string]any{{"title": "Data", "url": "https://example.invalid/d", "snippet": "sn"}},
	})
	if hits := parseSearchHits(string(data)); len(hits) != 1 {
		t.Fatalf("data hits = %v", hits)
	}
	if parseSearchHits("not-json") != nil {
		t.Fatal("garbage is not a hit list")
	}
	if parseSearchHits(`{"message":"ok"}`) != nil {
		t.Fatal("an object without rows is not a hit list")
	}
}

func TestAsTextAndAsIntCoverTheScalarCases(t *testing.T) {
	if asText(3.0) != "3" || asText(1.5) != "1.5" {
		t.Fatalf("number asText: %q %q", asText(3.0), asText(1.5))
	}
	if asText(true) != "true" || asText(false) != "false" || asText(struct{}{}) != "" {
		t.Fatal("bool/default asText")
	}
	if n, ok := asInt(float64(7)); !ok || n != 7 {
		t.Fatalf("float asInt %v %v", n, ok)
	}
	if n, ok := asInt(json.Number("8")); !ok || n != 8 {
		t.Fatalf("json.Number asInt %v %v", n, ok)
	}
	if n, ok := asInt("9"); !ok || n != 9 {
		t.Fatalf("string asInt %v %v", n, ok)
	}
	if _, ok := asInt("nope"); ok {
		t.Fatal("garbage asInt")
	}
	if n, ok := asInt(json.Number("x")); ok || n != 0 {
		t.Fatalf("bad json.Number asInt %v %v", n, ok)
	}
	if _, ok := asInt(true); ok {
		t.Fatal("bool asInt")
	}
	if firstText(map[string]any{"a": ""}, "a", "b") != "" {
		t.Fatal("firstText should miss empty keys")
	}
}
